File: app/src/main/java/com/example/mvvm_artichecture_sample/ui/adpater/MainAdapter.java
package com.example.mvvm_artichecture_sample.ui.adpater;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mvvm_artichecture_sample.R;
import com.example.mvvm_artichecture_sample.data.remote.apimodel.CountryModel;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class MainAdapter extends RecyclerView.Adapter<MainAdapter.MainViewHolder> implements View.OnClickListener {
	
	private Context mContext;
	private List<CountryModel> mList;
	private WeakReference<MainAdapterHanlder> hanlderWeakReference;
	
	public MainAdapter(Context context, MainAdapterHanlder hanlder) {
		mContext = context;
		hanlderWeakReference = new WeakReference<>(hanlder);
		mList = new ArrayList<>();
	}
	
	public void addItem(List<CountryModel> list) {
		mList.addAll(list);
		notifyDataSetChanged();
	}
	
	@NonNull
	@Override
	public MainViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(mContext).inflate(R.layout.main_item_list, parent, false);
		return new MainViewHolder(view);
	}
	
	@Override
	public void onBindViewHolder(@NonNull MainViewHolder holder, int position) {
		CountryModel model = mList.get(position);
		holder.title.setText(model.getName());
	}
	
	@Override
	public int getItemCount() {
		return mList != null ? mList.size() : 0;
	}
	
	@Override
	public void onClick(View v) {
		MainViewHolder holder = (MainViewHolder) v.getTag();
		int pos = holder.getLayoutPosition();
		clicked(pos);
	}
	
	private void clicked(int position) {
		MainAdapterHanlder handler = hanlderWeakReference.get();
		if (handler != null)
			handler.onItemClick(position);
	}
	
	class MainViewHolder extends RecyclerView.ViewHolder {
		TextView title;
		
		public MainViewHolder(@NonNull View itemView) {
			super(itemView);
			title = itemView.findViewById(R.id.title);
			
			itemView.setTag(this);
			itemView.setOnClickListener(MainAdapter.this);
		}
	}
	
	
}

File: app/src/main/java/com/example/mvvm_artichecture_sample/ui/adpater/MainAdapter.kt
package com.example.mvvm_artichecture_sample.ui.adpater

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.mvvm_artichecture_sample.R
import com.example.mvvm_artichecture_sample.data.remote.apimodel.Country
import java.util.*

class MainAdapter(private val mContext: Context, private val hanlder: MainAdapterHanlder) : RecyclerView.Adapter<MainAdapter.MainViewHolder>(), View.OnClickListener {
    private val mList: MutableList<Country>?

    init {
        mList = ArrayList()
    }

    fun addItem(list: List<Country>) {
        mList!!.addAll(list)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MainViewHolder {
        val view = LayoutInflater.from(mContext).inflate(R.layout.main_item_list, parent, false)
        return MainViewHolder(view)
    }

    override fun onBindViewHolder(holder: MainViewHolder, position: Int) {
        val model = mList!![position]
        holder.title.text = model.name
    }

    override fun getItemCount(): Int {
        return mList?.size ?: 0
    }

    override fun onClick(v: View) {
        val holder = v.tag as MainViewHolder
        val pos = holder.layoutPosition
        clicked(pos)
    }

    private fun clicked(position: Int) {
        hanlder.onItemClick(position)
    }

    inner class MainViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var title: TextView

        init {
            title = itemView.findViewById(R.id.title)

            itemView.tag = this
            itemView.setOnClickListener(this@MainAdapter)
        }
    }


}
